### PR TITLE
Update lighttpd.conf.debian

### PR DESCRIPTION
**Pull request by jauderho**
This addresses two issues. 

1) Restricting allowed access methods to just GET, POST and HEAD. Disallow HTTP OPTIONS method. See https://www.rapid7.com/db/vulnerabilities/http-options-method-enabled

2) Explicitly disable "url-path-2f-decode". Flagged as CVE-2019-11072

These popped up during a vulnerability scan.

**By submitting this pull request, I confirm the following:** 
*please fill any appropriate checkboxes, e.g: [X]*

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [X] I have made only one major change in my proposed changes.
- [X] I have commented my proposed changes within the code.
- [X] I have tested my proposed changes, and have included unit tests where possible.
- [X] I am willing to help maintain this change if there are issues with it later.
- [X] I give this submission freely and claim no ownership.
- [X] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [X] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

Please make sure you [Sign Off](https://github.com/pi-hole/pi-hole/wiki/How-to-signoff-your-commits.) all commits. Pi-hole enforces the [DCO](https://github.com/pi-hole/pi-hole/wiki/Contributing-to-the-project).

---
**What does this PR aim to accomplish?:**
Fix some identified vulnerabilities in lighttpd config

**How does this PR accomplish the above?:**
Added config items to address issues.

**What documentation changes (if any) are needed to support this PR?:**
See https://www.rapid7.com/db/vulnerabilities/http-options-method-enabled and CVE-2019-11072

---
* You must follow the template instructions. Failure to do so will result in your pull request being closed.
* Please respect that Pi-hole is developed by volunteers, who can only reply in their spare time.

### CODE REVIEW

**Comment by dschaper** [comment]
Does it work for all the supported versions of `lighttpd`?

**Comment by jauderho** [comment]
Tested on Raspian and multiple versions of Debian. Seem to work fine.

**Comment by dschaper** [comment]
Just to make sure, Raspbian Stretch and Raspbian Buster?

**Comment by MichaIng** [comment]
This does not work in Debian/Raspbian Stretch. Those options are part of Lighttpd 1.4.53: https://www.lighttpd.net/2019/1/27/1.4.53/
Raspbian/Debian Stretch ship Lighttpd 1.4.45: https://packages.debian.org/stretch/lighttpd

Moreover these URL normalisation causes issues when using GET requests within PHP: https://github.com/nextcloud/server/issues/17065#issuecomment-573243754
This leads to Lighttpd decoding URL encodings. But PHP decodes this a second time, automatically for the `_GET` array. IMO it is not job of the webserver to manipulate the URL and especially not the query string, as it cannot know what the intention was.

Practically, from what we found, the only affected problematic character is `+`, which is regularly used for access tokens. So in the particular Nextcloud case:
- On logouts (and in some other cases), the logout script is called via GET request with the access token as encoded query string.
- Lighttpd with `url-normalize-required` enabled now decodes the query string, which leads to the encoded `+` (`%2B`) being decoded to `+`, before the request is forwarded to PHP.
- Now PHP decodes the query string a second time before storing it in `_GET`. And since `+` is as well a URL encoding for white space, it is decoded to white space which leads to a wrong access token and hence a failing logout.

I remember that exactly the same was already an issue with Pi-hole was well when the default `lighttpd.conf` was used on Debian Buster. I'll see if I find this...

**Comment by gstrauss** [comment]
@MichaIng wrote
> Practically, from what we found, the only affected problematic character is +, which is regularly used for access tokens.
The issue with '+' was a bug and was fixed in lighttpd 1.4.55 (released Jan 2020) as soon as it was reported upstream.  (I wish people reported issues to the lighttpd maintainers!)  https://redmine.lighttpd.net/issues/2999

Unfortunately, Debian is extremely slow picking up patches.  Ongoing many-months-long attempt to update buster-backports:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=961843

**Comment by MichaIng** [comment]
> The issue with '+' was a bug and was fixed in lighttpd 1.4.55

Nothing we can rely on since we cannot know which package/version from which repo user has. And even 1.4.53 will not reach Debian Stretch which is officially supported by Pi-hole, hence these settings would then need to be added/removed conditionally.

Moreover IMO this was no "fix" but a work around the most urgent problem this setting ships, while potentially causing new problems. There are a few URL codes now not decoded (a few others besides "+") while most others are. Consequently this leads to half-decoded query strings, completely unreliable/unpredictable. Even when only taking the "+", which in fact codes for white space, having everything around it decoded but the "+" now not, while it cannot be known if the "+" was wanted as "+" or indeed as white space. In the end it cannot be derived anymore from any application which wants to parse the query string.

I never really understood what the purpose of these settings is. If there is any way to get unintended access or crash the server via URI, then this should be fixed internally and not by manipulating the URI (before passing it further). And for CVE-2019-11072 this is the case according to Lighttpd dev, if I understood correctly: https://www.rapid7.com/db/vulnerabilities/http-lighttpd-cve-2019-11072
> The developer states "The feature which can be abused to cause the crash is a new feature in lighttpd 1.4.50, and is not enabled by default. It must be explicitly configured in the config file (e.g. lighttpd.conf). Certain input will trigger an abort() in lighttpd when that feature is enabled. lighttpd detects the underflow or realloc() will fail (in both 32-bit and 64-bit executables), also detected in lighttpd. Either triggers an explicit abort() by lighttpd. This is not exploitable beyond triggering the explicit abort() with subsequent application exit."

I stay with my opinion that it cannot be the job ob the webserver to manipulate any single part of the URI, especially not the query string, as long as there are no explicit redirects/rewrites configured, at least as long as there is an external handler behind it (PHP).

**Comment by gstrauss** [comment]
(multi-posting to https://github.com/nextcloud/server/issues/17065)

> One never knows which part of the "unnormal" URI was intended and which one not, especially when its about the query string, but also thinking of special (intended) rewrites and such. Having now a few URL-codings decoded or changed while others stay the same only creates a completely unpredictable result for what any handler gets.

I think you misunderstood what is meant by **normalization**.  %-encoding has some complexity since there are slightly different rules for required characters to %-encode in different parts of the URI, but there is a normal form which consists of %-encoding the characters required to be %-encoded.  Then, if a specific application is trying encode information in the URI, and needs non-required characters to remain %-encoded to preserve semantics, an application can do so more robustly by using base64 or hex (without `%`) or other encoding.  If %-encoding is used instead, lighttpd provides some flags to loosen the normalization, e.g. `url-path-2f-decode`, `url-path-dotseg-remove`, `url-query-20-plus`.

>> If you want to protect your /admin page, you do not want someone to bypass the restriction with /%61dmin.

That example was oversimplified and does not apply to lighttpd.  lighttpd %-decodes the url-path component for use in `$HTTP["url"]` conditions in lighttpd.conf.  `..` path segments are also removed to defend against path traversal attacks (applies to all web servers).  This typically matches the user expectation of applying the lighttpd.conf condition to a path which often maps to a filesystem path.

However, URL rewrite and URL redirect produces a URI which must be properly %-encoded.  The matching rules constructed by the end-user should match what is intended, and the resulting URI constructed from captured matches should be what is intended by the end-user.  mod_rewrite and mod_redirect match on the %-encoded URI.  To do otherwise could easily lose information in the encoding and result in broken URIs.

URI normalization allows an end-user to write mod_rewrite and mod_redirect matching rules relying on the fact that the URI against which it is matching is in a normalized form.  When **you** disable this protection in lighttpd, **you** expose end-users to malicious actors that can manipulate the rewrite and redirect rules.

For those who need to construct complex URIs, lighttpd provides an enhanced syntax for %-encoding or %-decoding matched parts of the URI.  https://redmine.lighttpd.net/projects/lighttpd/wiki/Docs_ModRewrite

The official lighttpd code base does not ship with mod_rewrite or mod_redirect loaded, and does not ship with any rewrite or redirect rules.

URI normalization enhances security for end-users employing URL rewrite and URL redirect and gives the end-user more power in controlling the construction of the resulting URI.

Any well-maintained linux distro is either shipping lighttpd 1.4.55, or has backported the lighttpd patch for the %2b decoding bug.

URI normalization in lighttpd first shipped in lighttpd 1.4.50 (Aug 2018) and was enabled by default in lighttpd 1.4.54 (May 2019).  The %2b query string decoding bug was reported 7 months later in Jan 2020, patched the same day, and released in lighttpd 1.4.55 (Jan 2020)  https://redmine.lighttpd.net/issues/2999

**Comment by gstrauss** [comment]
@Michalng
> I never really understood what the purpose of these settings is.

I hope the above helps explain.

As for this pull-request, I would suggest omitting `server.http-parseopts`.

`server.http-parseopts` does not exist earlier than lighttpd 1.4.50.
For lighttpd 1.4.50 - lighttpd 1.4.53, `server.http-parseopts` is not enabled by default in lighttpd.  The settings suggested by @jauderho, including `"url-path-2f-decode" => "disable",` addresses CVE-2019-11072 for lighttpd 1.4.50 - 1.4.53, but this is not necessary in lighttpd 1.4.54 and later.  However, enabling `server.http-parseopts` exposes the %2b query string decoding bug, which was fixed in lighttpd 1.4.55.

tl;dr:
* omit `server.http-parseopts` and use lighttpd defaults, which enable `server.http-parseopts` by default in lighttpd 1.4.54, *after*  CVE-2019-11072 was fixed.
* If someone is running lighttpd 1.4.54 and encounters the %2b query string decoding bug, suggest upgrading to lighttpd 1.4.55.

(Aside: while lighttpd did not enable `server.http-parseopts` by default until lighttpd 1.4.54, Debian Buster ships with lighttpd 1.4.53 and the Debian default lighttpd.conf enabled `server.http-parseopts` in lighttpd 1.4.53 shipped with Debian Buster.  This will eventually be fixed in Debian Buster with https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=961843)

**Comment by MichaIng** [comment]
In principle I'm still not convinced. I mean of course when its about `$HTTP["url"]` rules, these must be applied after all path decoding has been done, so directly before the actual resolved resource would be accessed. If its a resource outside of the webroot (through `..`, which should otherwise work, it does perfectly fine on Apache2!) then 403 or rewrite to webroot, excluding symlinks as those must be expected as intended, otherwise there are usually options to disable following symlinks, which disables rewrites usually as well due to the similar behaviour. I think Lighttpd behaves like this without having any of these options enabled, but probably find time to verify.
But when the resource is identified and it is a handler outside of Lighttpd, I'd always expect the original raw URL to be passed and leave any interpretation to the external handler. And the query string is something that Lighttpd anyway never handles directly (?) hence especially this must never be touched.
But I'm no webserver specialist, hence maybe the above idea about how things should or do usually work is wrong.

> That example was oversimplified and does not apply to lighttpd.

Since this is all about Lighttpd (never saw similar options for other webservers or related issues or security risks), do you know any concrete example that applies to Lighttpd?

> When you disable this protection in lighttpd, you expose end-users to malicious actors that can manipulate the rewrite and redirect rules.

As above, any concrete example that does not imply fragile configs, like the aim to block access to resources via [rewrite_raw](https://doc.lighttpd.net/lighttpd2/mod_rewrite.html) instead of `url.access-deny`?

Don't understand me wrong, I really try to understand a real case where disabling this options would expose a security/privacy issue that cannot be reasonably fixed either Lighttpd internally or the external handler.

@ Pi-hole team: Btw sorry for driving this discussion here, I hope its in your interest as well. I faced issues with these options in two cases already and never with any other webserver, so I really try to understand why Lighttpd is (claimed to be) a special case here.

> omit server.http-parseopts and use lighttpd defaults, which enable server.http-parseopts by default in lighttpd 1.4.54, after CVE-2019-11072 was fixed.

That is the case currently with Pi-hole and as I read as well that (some of) those options are enabled by default I mentioned at some other place that this is a topic Pi-hole needs to take care of or at least keep in mind when facing any issues with GET requests. But as of my last tests actually those options were NOT enabled by default when omitted, probably the statement in the changelogs was about the default (upstream) `lighttpd.conf` which has those set and enabled?

> If someone is running lighttpd 1.4.54 and encounters the %2b query string decoding bug, suggest upgrading to lighttpd 1.4.55.

IMO not a suitable solution. Not everyone is able or wants to deal with compiling the newest version from sources or even adding backports (not available on Raspbian!) and such. If a certain setting causes an issue, it should be either disabled by default or conditionally based on derived Lighttpd version, to have an OOTB running setup for all users on all official supported distros with standard package sources.
___________
So left is `request-method` limit which seems okay although adds zero security benefit with the apparent config. Default is GET, POST, OPTIONS and HEAD which is all that the OPTIONS method tells you which any attacker can expect without doing the OPTIONS request since its default on all webservers. There might be a reason when you allow PUT, DELETE and such, although it can be simply tested by sending each of those requests and see if you get a 405 (method not allowed) or not. But anyway its security through obscurity topic and one better takes care to that all methods require secure authentication instead feeling secure by hiding those. But since the Pi-hole lighttpd.conf is meant to make (only) Pi-hole working, I don't see an issue with explicitly only allowing the methods required for it to run.

**Comment by gstrauss** [comment]
@MichaIng the link you provided to rewrite_raw is to lighttpd2 documentation and is out of scope.  Please correct or remove the link.  lighttpd2 has never been released and is not under active development.  lighttpd1.4 is under active development.

> do you know any concrete example that applies to Lighttpd?

> As above, any concrete example that does not imply fragile configs, like the aim to block access to resources via rewrite_raw instead of url.access-deny?

I know of concrete examples which apply to multiple webservers.  I already pointed to the interaction with rewrite and redirect rules, but will not detail further here.  As you suggested, issues can arise in fragile configs on any web server.

> >  If someone is running lighttpd 1.4.54 and encounters the %2b query string decoding bug, suggest upgrading to lighttpd 1.4.55.

> IMO not a suitable solution. Not everyone is able or wants to deal with compiling the newest version from sources or even adding backports (not available on Raspbian!) and such. If a certain setting causes an issue, it should be either disabled by default or conditionally based on derived Lighttpd version, to have an OOTB running setup for all users on all officially supported distros with standard package sources.

Why not?  Which distros are running lighttpd 1.4.54?  Debian Buster is on 1.4.53, and buster-backports has lighttpd 1.4.55.  Most other popular distributions are running lighttpd 1.4.55.  To be running lighttpd 1.4.54, it is likely you built it yourself, or that your distro has lighttpd 1.4.55 available and you have not yet upgraded.

> But I'm no webserver specialist, hence maybe the above idea about how things should or do usually work is wrong.

You have stated many strong opinions, as well as the contradictory statement that you are not a webserver specialist.  You have made more than a few questionable assumptions, but it is off-topic for me to try to point them all out.

OTOH, I hope that you will re-read and reconsider my statements as speaking with expertise (though not infallibility) as I am a lighttpd 1.4 developer, and I wrote `server.http-parseopts` and all of the lighttpd code being discussed, including both the bugs and the fixes.

tl;dr: A simple lighttpd config is likely more secure than a complex one.  Please use the lighttpd defaults unless you specifically need to tune the behavior.

**Comment by dschaper** [comment]
@gstrauss 

Since we haven't heard from the OP, would you mind setting up a PR based on the `development` branch with the changes you feel are the proper ones? Instead of spending time speculating what may or may not work, let's just put things to in vivo tests.

**Comment by gstrauss** [comment]
My recommendation is to omit `server.http-parseopts` from this PR.
The remaining part of OP's patch is the following, which IMO is minor, and up to pi-hole developers to choose if the additional restriction is useful to pi-hole.
```
$HTTP["request-method"] !~ "^(GET|POST|HEAD)$" {
	url.access-deny = ("")
}
```

**Comment by dschaper** [comment]
Thanks.

I think adding more restrictions is not the proper way to go. Hard coding in that directive will be hard/impossible for users to override with external configurations if they so want to. It's easier for the users that want this to enable the directive with their own external configuration.

**Comment by MichaIng** [comment]
> I wrote server.http-parseopts and all of the lighttpd code being discussed, including both the bugs and the fixes.

Ah that is useful information, may I ask you to clarify two things:
- With Lighttpd 1.4.53 `url-normalize-required` just needs to be removed/commented/skipped to fix the "+" case, since it is not internally enabled by default?
- Since Lighttpd 1.4.54 one would need to actively set it to disabled since it is internally enabled by default?

And as a suggestion, since I as well saw https://redmine.lighttpd.net/issues/2999 when facing the bug in Nextcloud the first time (hence not reporting it again) and I was quite hard annoyed since a large amount of time went into debugging and testing to finally see that Lighttpd is manipulating an encrypted token in the query string, which really never should happen and no one was even close to expecting and which is why I am still sceptic about the whole concept and made such strong statements above, all from the PHP developers perspective which exposed their users to a major security issue due to this (thing about users staying logged in on public computers), I hope you understand. Would it be possible and much more reasonable to only normalise everything until+including the path element and leave the query string and fragment untouched? Both is (if I am not mistaken) irrelevant for the discussed security aspects but is regularly used for tokens/strings which in question must remain 100% what they were. It is probably "lucky" if the strings have "+" as most special character, but in theory others can be used, like "%" itself. I see that some options are explicitly for the query string, but that is really a dangerous thing (even mentioned in lighttpd.conf) and there I see really zero security aspect. Lighttpd itself uses the query string only for sorting of directory indexes, or is there anything else? The major problem is that as mentioned before no other webserver does such, hence no PHP developer is expecting it, and non-matching query strings can lead to all thinkable smaller and larger problems, sometimes not directly visible to the end user, which is even worse. Some users tend to enable everything that is shown by any random guide in the internet as "security" or "privacy" option, then running into issues they cannot associate (we just had the referrer=none case in Pi-hole) and I fear that those Lighttpd options are a similar case with in mentioned cases the radically opposite result.

> Why not? Which distros are running lighttpd 1.4.54? Debian Buster is on 1.4.53, and buster-backports has lighttpd 1.4.55. Most other popular distributions are running lighttpd 1.4.55. To be running lighttpd 1.4.54, it is likely you built it yourself, or that your distro has lighttpd 1.4.55 available and you have not yet upgraded.

I can confirm that 1.4.54 is no shipped by any official Debian or Ubuntu repo, not checked for rpm distros.

**Comment by dschaper** [comment]
Can you move this to somewhere more appropriate please?

**Comment by gstrauss** [comment]
I think this PR can be closed.  @dschaper opted above not to modify `server.http-parseopts`

Even so, patches to lighttpd (released Jan 2020) are (finally) available in Debian buster:
26 Sep 2020, Debian finally backported lighttpd patches to Debian buster
bug closed: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=961843
Released 26 Sep 2020 in Debian Buster 10.6

Debian buster-backports has had lighttpd 1.4.55 available since early Mar 2020.

@MichaIng noted in https://github.com/nextcloud/server/issues/17065
Raspbian uses Debian sources, only build for armv6hf. The fixed version is available there as well: http://raspbian.raspberrypi.org/raspbian/pool/main/l/lighttpd/
